src/engines/evaluate-join-potential.ts

```diff
@@ -0,0 +1,71 @@
+import { FastNode } from './FastNode'
+import { propagateJoinMessages } from './join-propagation'
+import { evaluate, evaluateMarginalPure } from './evaluation'
+import { FastClique } from './FastClique'
+import { Formula } from './Formula'
+import { FastPotential } from './FastPotential'
+import { sum } from 'ramda'
+
+/** Construct the join of an arbitrary collection of variables in a Bayesian Network,
+ * conditioned on an optional set of parent variables.  Returns the potential function
+ * of the joint distribution over the head and parent variables subject to any evidence.
+ * @param nodes: The collection of nodes in the bayesian network
+ * @param cliques: The collection of cliques in the junction tree for the bayesian
+ *   network.
+ * @param formulas: The collection of formulas for computing the posterior
+ *   distributions for the cliques, nodes and separators  of the Bayes Network, and
+ *   all the intermediate potentials.
+ * @param potentialFunction: The collection of potential functions for the cliques
+ *   nodes and separators of the Bayes Network and the intermediate potentials.
+ * @param headVariables: The indices of the variables that occur as head variables of
+ *   the joint distribution being constructed.  These indices must be valid references
+ *   to variables in the network.
+ * @param parentVariables: The idices of the variables that occur as parent variables
+ *   of the joint distribution being constructed.   These indices must be valid
+ *   references for the variables in the network, and must be distinct from the
+ *   head variables of the joint being constructed.
+ * @returns: The potential function for the join which satisfies the conditions
+ *   that for every combination of parents, the potentials sum to unity.  The
+ *   case when there are no parents (unconditioned joint distribution), corresponds
+ *   to the trivial case where there is only a single (empty) combination of
+ *   parent values.
+ */
+export function evaluateJoinPotentials (nodes: FastNode[], cliques: FastClique[], connectedComponents: number[][], separators: number[][], formulas: Formula[], potentials: (FastPotential | null)[], headVariables: number[], parentVariables: number[]): { potentials: FastPotential; supplementalFormulas: Formula[]; supplementalPotentials: (FastPotential | null)[]} {
```

Comment:
How long are the lines formatted to be?  This seems like some odd line wrapping.

Reply:
The upstream repository has no limitation on the line length.   I prefer to leave the type spec as a single line, since the documentation is arranged in a nicer format for reading.